lib/gc/Transforms/IterativeTilingAndFusion.cpp

```diff
@@ -0,0 +1,853 @@
+      if (llvm::detail::isPresent(sliceOp))
+        return failure();
+      sliceOp =
+          dyn_cast<OffsetSizeAndStrideOpInterface>(useOfResult.getOwner());
+    } else if (auto yieldOp = dyn_cast<scf::YieldOp>(useOfResult.getOwner())) {
+      if (auto loop = dyn_cast<LoopLikeOpInterface>(yieldOp->getParentOp()))
+        return getClosestInsertSliceOfResult(
+            loop->getResult(useOfResult.getOperandNumber()));
+    }
+  }
+
+  if (!llvm::detail::isPresent(sliceOp))
+    return failure();
+  else {
+    return sliceOp;
+  }
+}
+
+struct CandidateDefOrUse {
+  enum Type { def = 0, use };
+  Operation *ownerOp;
+  Type type;
+  union {
+    OpOperand *operand;
+    OpResult result;
+  };
+
+  CandidateDefOrUse(OpResult resultOfDefOp)
+      : ownerOp(resultOfDefOp.getDefiningOp()), type(Type::def),
+        result(resultOfDefOp) {}
+  CandidateDefOrUse(OpOperand *operandOfUseOp)
+      : ownerOp(operandOfUseOp->getOwner()), type(Type::use),
+        operand(operandOfUseOp) {}
+
+  bool isDef() const { return type == Type::def; }
+  bool isUse() const { return type == Type::use; }
+};
+
+using CandidateSliceFilter = std::function<LogicalResult(
+    RewriterBase &, OffsetSizeAndStrideOpInterface, CandidateDefOrUse)>;
+
+using CandidateSliceComparer =
+    std::function<int(RewriterBase &, OffsetSizeAndStrideOpInterface,
+                      OffsetSizeAndStrideOpInterface, CandidateDefOrUse)>;
+
+static LogicalResult
+noTilingOnReductionFilter(RewriterBase &rewriter,
+                          OffsetSizeAndStrideOpInterface candidate,
+                          CandidateDefOrUse defOrUse) {
+  linalg::LinalgOp linalgOp = dyn_cast<linalg::LinalgOp>(defOrUse.ownerOp);
+  if (!linalgOp)
+    return success();
+
+  AffineMap affMap =
+      defOrUse.isDef() ? linalgOp.getIndexingMapMatchingResult(defOrUse.result)
+                       : linalgOp.getMatchingIndexingMap(defOrUse.operand);
+
+  TilingInterface tilableOp = dyn_cast<TilingInterface>(defOrUse.ownerOp);
+  SmallVector<Range> iterDomain = tilableOp.getIterationDomain(rewriter);
+  SmallVector<utils::IteratorType> iterTypes = tilableOp.getLoopIteratorTypes();
+  SmallVector<OpFoldResult> tileSizes = candidate.getMixedSizes();
+  // check reduction iteration is full on TileSizes
+  for (const auto &resultExpr : llvm::enumerate(affMap.getResults())) {
+    unsigned iterPosition =
+        cast<AffineDimExpr>(resultExpr.value()).getPosition();
+    if (iterTypes[iterPosition] == utils::IteratorType::reduction) {
+      std::optional<int64_t> cstIterDomain =
+          getConstantIntValue(iterDomain[iterPosition].size);
+      FailureOr<int64_t> cstTileSizes =
+          ValueBoundsConstraintSet::computeConstantBound(
+              presburger::BoundType::UB, tileSizes[resultExpr.index()], nullptr,
+              true);
+      if (!cstIterDomain || failed(cstTileSizes) ||
+          cstIterDomain != cstTileSizes)
```

Comment:
As the name of this filter(`noTilingOnReductionFilter`) said, we need to ensure there exist no tiling on any reduction dimension. Otherwise, it will lead to wrong calculation result. I.e. `tileSize` should equal to `IterationDomain`. Certainly, if either of them is not dynamic, they are incomparable until RUNTIME , so just return `failure` for such case.